src/utils.rs

```diff
@@ -0,0 +1,92 @@
+//! Low-level utility functions.
+
+use core::{
+    any::{type_name, TypeId},
+    mem, ptr,
+};
+
+/// Determine if two static, generic types are equal to each other.
+#[inline(always)]
+pub(crate) fn type_eq<T: 'static, U: 'static>() -> bool {
+    // Reduce the chance of `TypeId` collisions causing a problem by also
```

Comment:
`TypeId` is by definition globally unique. To be sure, collisions [*can* currently occur](https://github.com/rust-lang/rust/issues/10389), but I still don’t think these additional checks are warranted, given the extreme rarity of the issue, that it’s acknowledged to be a soundness hole in Rust, that [a fix (more or less) is in the pipeline](https://github.com/rust-lang/rust/pull/95845), and that the additional checks here still don’t catch everything.

Reply:
Fair enough, though none of these checks show up in a final binary since they are trivially optimized away so it doesn't really hurt. (As an aside, these checks were already present before this PR, they're just being rearranged here.)